Tests/SwiftAlgorithmsTests/TestUtilities.swift

```diff
@@ -66,3 +66,192 @@ func XCTAssertEqualSequences<S1: Sequence, S2: Sequence>(
 }
 
 func XCTAssertLazy<S: LazySequenceProtocol>(_: S) {}
+
+/// Tests that all index traversal methods behave as expected.
+///
+/// Verifies the correctness of the implementations of `startIndex`, `endIndex`,
+/// `indices`, `count`, `isEmpty`, `index(before:)`, `index(after:)`,
+/// `index(_:offsetBy:)`, `index(_:offsetBy:limitedBy:)`, and
+/// `distance(from:to:)` by calling them with just about all possible input
+/// combinations. When provided, the `indices` function is used to to test the
+/// collection methods against.
+///
+/// - Parameters:
+///   - collections: The collections to be validated.
+///   - indices: A closure that returns the expected indices of the given
+///     collection, including its `endIndex`, in ascending order. Only use this
+///     parameter if you are able to compute the indices of the collection
+///     independently of the `Collection` conformance, e.g. by using the
+///     contents of the collection directly.
+///
+/// - Complexity: O(*n*^3) for each collection, where *n* is the length of the
+///   collection.
+func validateIndexTraversals<C>(
+  _ collections: C...,
+  indices: ((C) -> [C.Index])? = nil,
+  file: StaticString = #file, line: UInt = #line
+) where C: BidirectionalCollection {
+  for c in collections {
+    let indices = indices?(c) ?? (c.indices + [c.endIndex])
```

Comment:
I find it a bit confusing that this shares the name of `c.indices` but has one more element — could this be `allIndexes` / `allValidIndices` / `indicesIncludingEnd`, etc?

Reply:
Great idea, I'll go with `indicesIncludingEnd`. This also makes the `- 1` on the next line less confusing.